src/gui/dialogs/preferences_dialog.cpp

```diff
@@ -1213,9 +1212,14 @@ void preferences_dialog::handle_theme_select()
 
 void preferences_dialog::handle_gui2_theme_select()
 {
-	find_widget<label>(get_window(), "restart_msg", false).set_visible(widget::visibility::visible);
 	menu_button& gui2_theme_list = find_widget<menu_button>(this, "choose_gui2_theme", false);
-	prefs::get().set_gui_theme(gui2_themes_.at(gui2_theme_list.get_value()));
+	unsigned selected_theme = gui2_theme_list.get_value();
+	if (selected_theme != current_gui_theme_) {
+		current_gui_theme_ = selected_theme;
+		prefs::get().set_gui_theme(gui2_themes_.at(selected_theme));
+		gui2::init(gui2_themes_.at(selected_theme), false);
```

Comment:
Would it be enough to call `guis[selected_theme].activate()`?

Reply:
No, because `activate()` is a bit of a misnomer and doesn't actually activate the theme. It actually sets some accessory variables.
https://github.com/wesnoth/wesnoth/blob/aba0abeb798465e62fecf2892802f9b3ded96183/src/gui/core/gui_definition.cpp#L146-L159

The actual line responsible for "activating" the theme is
https://github.com/wesnoth/wesnoth/blob/aba0abeb798465e62fecf2892802f9b3ded96183/src/gui/gui.cpp#L81

I have some simplifications in mind for `gui` and `gui_definition` though, which might make it exactly that.

Reply:
What I'm thinking is we don't need to keep the itors here, we just need the corresponding id.
https://github.com/wesnoth/wesnoth/blob/aba0abeb798465e62fecf2892802f9b3ded96183/src/gui/core/gui_definition.cpp#L30-L31

Which will simplify https://github.com/wesnoth/wesnoth/blob/aba0abeb798465e62fecf2892802f9b3ded96183/src/gui/gui.cpp#L69-L92 a lot.

Reply:
I assume the iterators are used since otherwise you have to search for the id each time.

So it should be enough then to set the iterator and call activate?

Reply:
> I assume the iterators are used since otherwise you have to search for the id each time.

will a hashtable (std::unorder_map) instead std::map work?

> So it should be enough then to set the iterator and call activate?

There's also:
https://github.com/wesnoth/wesnoth/blob/aba0abeb798465e62fecf2892802f9b3ded96183/src/gui/gui.cpp#L44-L45

Reply:
Using unordered_map still means you need to hash the ID every time, though that's probably marginally faster than searching a balanced binary tree every time.

Ultimately I don't think the choice matters much at all – we're not going to have a billion themes installed.

Reply:
I don't think performance is a major concern it's just odd to do unnecessary steps. What's the issue with the iterator?